Src/GoogleApis.Core/Apis/Util/PageStreamer.cs

```diff
@@ -0,0 +1,137 @@
+﻿// Copyright 2015 Google Inc. All Rights Reserved.
+// Licensed under the Apache License Version 2.0.
+
+using System;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Google.Apis.Util
+{
+    /// <summary>
+    /// A page streamer is a helper to provide both synchronous and asynchronous page streaming
+    /// of a listable or queryable resource.
+    /// </summary>
+    /// <remarks>
+    /// <para>
+    /// The expected usage pattern is to create a single paginator for a resource collection,
+    /// and then use the instance methods to obtain paginated results. The methods take parameters
+    /// expressing how individual pages are fetched, allowing the same paginator to be used across
+    /// multiple contexts (for example, with different client instances).
+    /// </para>
+    /// </remarks>
+    /// <typeparam name="TResource">The type of resource being paginated</typeparam>
+    /// <typeparam name="TRequest">The type of request used to fetch pages</typeparam>
+    /// <typeparam name="TResponse">The type of response obtained when fetching pages</typeparam>
+    /// <typeparam name="TToken">The type of the "next page token", which must be equatable to itself</typeparam>
+    public sealed class PageStreamer<TResource, TRequest, TResponse, TToken>
+        where TToken : IEquatable<TToken>
+    {
+        private readonly Func<TRequest, TToken, TRequest> _requestProvider;
+        private readonly Func<TResponse, TToken> _tokenExtractor;
+        private readonly Func<TResponse, IEnumerable<TResource>> _resourceExtractor;
+        private readonly TToken _emptyToken;
```

Comment:
I don't know; there's no reason it *couldn't* be (say) an int with a value of 0 for "done". I didn't want to bake too many assumptions in. In gRPC it's more likely to be `""` or `ByteString.Empty`, but of course that's not too relevant for this code.